src/relay/pass/type_infer.cc

```diff
@@ -390,6 +390,23 @@ class TypeInferencer : private ExprFunctor<Type(const Expr&)> {
     auto ret = FuncTypeNode::make(arg_types, rtype, f->type_params, {});
     return solver_.Resolve(ret);
   }
+
+  Type VisitExpr_(const RefNewNode* op) final {
+    return RefTypeNode::make(GetType(op->value));
+  }
+
+  Type VisitExpr_(const RefReadNode* op) final {
+    Type it = IncompleteTypeNode::make(TypeVarNode::Kind::kType);
+    this->Unify(GetType(op->ref), RefTypeNode::make(it), op->span);
+    return it;
```

Comment:
Why unify with a fresh incomplete type? You could just return `GetType(op->ref)`

Reply:
this propagate the constraint backward (it unify op->ref with a reftype

Reply:
Ah, I misread the second line. Thanks for clarifying